misk/src/test/kotlin/misk/perf/PauseDetectorTest.kt

```diff
@@ -0,0 +1,128 @@
+    val (pauseTimeMillis, shortestObservedDeltaNsec) = detector.check()
+    assertThat(pauseTimeMillis).isEqualTo(0)
+    verifyShortestObservedDelta(shortestObservedDeltaNsec)
+    assertThat(takeLogs()).isEmpty()
+  }
+
+  private fun verifyShortestObservedDelta(shortestObservedDeltaNsec: Long) {
+    assertThat(shortestObservedDeltaNsec).isEqualTo(TimeUnit.MILLISECONDS.toNanos(1))
+  }
+
+  @Test fun `no logging`() {
+    detector.sleep()
+    fakeTicker.sleepMs(99)
+    detector.check()
+    takeLogs().isEmpty()
+  }
+
+  @Test fun `log info on a pause`() {
+    detector.sleep()
+    fakeTicker.sleepMs(101)
+    val (pauseTimeMillis, shortestObservedDeltaNsec) = detector.check()
+    assertThat(pauseTimeMillis).isEqualTo(101)
+    verifyShortestObservedDelta(shortestObservedDeltaNsec)
+    val logs = takeLogs()
+    assertThat(logs).hasSize(1)
+    assertThat(logs.first().message).isEqualTo("Detected JVM pause of 101 ms")
+    assertThat(logs.first().level).isEqualTo(Level.INFO)
+  }
+
+  @Test fun `log warn on a pause`() {
+    detector.sleep()
+    fakeTicker.sleepMs(2322)
+    val (pauseTimeMillis, shortestObservedDeltaNsec) = detector.check()
+    assertThat(pauseTimeMillis).isEqualTo(2322)
+    verifyShortestObservedDelta(shortestObservedDeltaNsec)
+    val logs = takeLogs()
+    assertThat(logs).hasSize(1)
+    assertThat(logs.first().message).isEqualTo("Detected JVM pause of 2322 ms")
+    assertThat(logs.first().level).isEqualTo(Level.WARN)
+  }
+
+  @Test fun `log error on a pause`() {
+    detector.sleep()
+    fakeTicker.sleepMs(99999)
+    val (pauseTimeMillis, shortestObservedDeltaNsec) = detector.check()
+    assertThat(pauseTimeMillis).isEqualTo(99999)
+    verifyShortestObservedDelta(shortestObservedDeltaNsec)
+    val logs = takeLogs()
+    assertThat(logs).hasSize(1)
+    assertThat(logs.first().message).isEqualTo("Detected JVM pause of 99999 ms")
+    assertThat(logs.first().level).isEqualTo(Level.ERROR)
+  }
+
+  @Test fun `ticker goes backwards`() {
+    // 1ms of sleep
+    detector.sleep()
+    // Move the ticker back 2020ms
+    fakeTicker.sleepMs(-2020)
+    val (pauseTimeMillis, shortestObservedDeltaNsec) = detector.check()
+    assertThat(pauseTimeMillis).isEqualTo(-2020L)
+    verifyShortestObservedDelta(shortestObservedDeltaNsec)
+    val logs = takeLogs()
+    assertThat(logs).hasSize(1)
+    assertThat(logs.first().message)
+      .isEqualTo("Observed a negative pause time of 2020ms. Non-monotonic ticker?")
+    assertThat(logs.first().level).isEqualTo(Level.INFO)
+  }
+
+  private fun takeLogs(): List<ILoggingEvent> = logCollector.takeEvents(PauseDetector::class)
+
+  class TestModule : KAbstractModule() {
+    override fun configure() {
+      // Wire up the detector
+      val config = PauseDetectorConfig(
+        resolutionMillis = 1,
+        logInfoMillis = 100,
+        logWarnMillis = 1000,
+        logErrorMillis = 10000
+      )
+      // NB: We are intentionally _not_ installing the module
```

Comment:
What does NB mean?

Reply:
https://en.wikipedia.org/wiki/Nota_bene

(This is a convention we've used on the Square side for a long time...and I see it has worked its way into Cash as well)